Travis CI: Stages, OS update, apt-get relief!

cmonr opened the pull request:
### Description

<!--
    Required
    Add here detailed changes summary, testing results, dependencies
    Good example: https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html (Pull request template)
-->
 
A rather large refactor of Travis CI. 

Changes:
- Addition of build stages.
  - Adds some much needed job organization
  - Runs jobs in batches, allowing for better parallelism
  - Allows for better control over job setup (ie: no need to install python modules when only running a shell command)
- Linux OS update.
  - All jobs are now running Ubuntu Xenial.
- Creation of bash helper functions.
  - Helper functions are located in `tools/test/travis-ci/functions.sh`
  - Functions can even be sourced and run locally (with caveats)
  - Cleans up Travis CI config (a bit)
- Misc changes
  - No more _apt-get update_! 🎉 
  - Enabled more caching
  - Some packages are now being sourced from S3
  - Easily updatable GCC archive when the time is right
  - Cleaned up some of the larger jobs

PR will be squashed and cleaned up tomorrow, along with the addition of Release Notes.

### Pull request type

<!--
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [x] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

<!--
    Optional
    Request additional reviewers with @username
-->

### Release Notes

<!--
    Optional
    In case of breaking changes, functionality changes or refactors, please add release notes here. 
    For more information, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing   /workflow.html#pull-request-types).
-->

Travis CI's testing config has been updated with some readability and performance improvements in mind.

#### Build Stages
Build stages have been added, which helps group similar jobs according to function.

Build stages also enocourage greater build parallelism by limiting the number of concurrent build jobs that a single PR can have, allowing multiple PRs to be tested in parallel instead of wauting in a build queue.

Finally, batching jobs into build stages allows for better job configuration. Some jobs that shouldn't have taken long were doing unneccesary actions such as downloading packages or installing python modules, which has now been corrected.

#### Python 3.7 by default
Jobs that require Python to run now use Py37 by default.

#### Ubutnu Xenial
The Linux VM that runs tests has been updated to use Xenail, Travis CI's latest supported Linux OS.

#### Minimal apt-get usage
Usage of `apt-get` has been greatly minimized by moving packages over to S3 and using `dpkg` to install cached packages when needed.

A major pain point would occur when a job would fail because the VM failed to run `apt-get update` successfully.

#### Bash functions
Common functions such as downloading packages or updating a job's status have now been collected into a single bash function. No strage shell invocations need to be done in order to use them which was not the case before.

0xc0170 commented on tools/test/travis-ci/functions.sh:
```diff
@@ -0,0 +1,139 @@
+#!/bin/bash -euf
```
license header file here

cmonr replied:
Added

0xc0170 reviewed (dismissed):
LGTM otherwise

Commit:
Changed default language to shell.

Not every VM needs the setup overhead that 'language: python' was providing

Commit:
Removed status reporting.

Re-enabled in later commit.

Commit:
Grouped jobs into build stages

Commit:
Re-enabled PR status reporting.
Added external bash file to source functions from

Commit:
Removed nearly all usage of 'apt-get'.

Primary cause of false Travis CI failures was running 'apt-get update'.
Refactored all instances where this was needed with manually fetching and installing dependencies in S3 instace.

This also includes GCC installation.

Commit:
Fixed littlefs tests to be Py3 compat

SeppoTakalo reviewed (approved):
I'm neutral regarding the changes as this lands probably OP's area to approve.
I don't see anything wrong, but Travis is not my expertise area anyway.


I hope you have agreed who will keep the .deb repository up to data.

bridadan commented on .travis.yml:
```diff
@@ -220,24 +241,27 @@ matrix:
+          #if [ "$PREV" -ne 0 ]
+          #then
+          #    STATUSM="$STATUSM ($(python -c "print '%+d' % ($CURR-$PREV)") cycles)"
+          #fi
```
Can we drop these comments?

cmonr replied:
Ah! That's what I missed.

I think these are still needed (@geky to comment), but I meant to update the bash variables and these jobs.

bridadan commented on .travis.yml:
```diff
@@ -291,38 +315,8 @@ matrix:
+          #if [ "$PREV" -ne 0 ]
+          #then
+          #    STATUSM="$STATUSM ($(python -c "print '%+.2f' % (100*($CURR-$PREV)/$PREV.0)")%)"
+          #fi
```
More comments to potentially remove

bridadan reviewed (changes requested):
Fantastic work @cmonr, seriously. This was sorely needed, hope this makes your life easier going forward!

Commit:
Cleaned up LittleFS and Event Queue tests.
Added {} around each variable.
Ported python invocations for simple math opts to coreutil commands

Commit:
Fixed shellcheck issues found in functions.sh
Added comments to functions

Commit:
Fixed status reporting.
'tr' command wasn't doing any replacement

Commit:
Made typo to unblock final check
